### PR TITLE
Fold in sml

### PR DESCRIPTION
**Pull request by mociarain**
An initial addition of the faculty-cli (sml) to the faculty library. This is just a wholesale copy and paste with some s/sherlock/faculty for good measure. i.e.

sml projects --> faculty projects

### CODE REVIEW

**Comment by srstevenson** [comment]
This is -1 from me: `sml` is legacy, untested code, which wasn't written for third party use. Additionally, it implements clients for all the Faculty services itself rather than using the tested clients in the `faculty` library. We should implement a new CLI with good test coverage and treating Windows as a supported platform, using the `faculty` library.

**Comment by mociarain** [comment]
Sorry I should have been clearer for people outside the engineering team. This is meant as an interim solution to get a faculty CLI. The duplication and lack of tests is not meant to be permanent or even that temporary. As for starting from scratch ¯\_(ツ)_/¯

File: faculty/cli/config.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            fg="yellow",
+        )
+    return profile_name
+
+
+def get_profile():
+    global _PROFILE_CACHE
+    if _PROFILE_CACHE is None:
+        profile_name = (
+            os.getenv("FACULTY_PROFILE")
```
Comment (acroz) [inline faculty/cli/config.py]: If support for `*_ENV` above is dropped, we won't need to check `*_PROFILE` here as we can just rely on `faculty.config.resolve_profile` to do that.

File: faculty/cli/config.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import os
+
+import click
+import faculty.config
```
Comment (acroz) [inline faculty/cli/config.py]: We usually put an empty line between third party (`click`) and package-internal (`faculty.config`) imports.

File: faculty/cli/client.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+from contextlib import contextmanager
+
+import requests
+import faculty.cli.auth
```
Comment (acroz) [inline faculty/cli/client.py]: We usually put an empty line between third party (`click`) and package-internal (`faculty.config`) imports.

File: faculty/cli/cli.py [hunk 4]
```diff
@@ -0,0 +1,1193 @@
...
+def _print_and_exit(msg, code):
+    """Print error message and exit with given code."""
+    click.echo(msg, err=True)
+    sys.exit(code)
+
+
+def _get_pypi_versions():
+    """List releases available from PyPI."""
+    # TODO needs to point to the new repo
+    response = requests.get("https://pypi.python.org/pypi/sml/json", timeout=1)
```
Comment (acroz) [inline faculty/cli/cli.py]:
```suggestion
    response = requests.get("https://pypi.org/pypi/faculty/json", timeout=1)
```

Comment (acroz) [reply faculty/cli/cli.py]: (Can also then remove the comment above)

File: faculty/cli/cli.py [hunk 5]
```diff
@@ -0,0 +1,1193 @@
...
+    versions = response.json()["releases"].keys()
+    return [StrictVersion(v) for v in versions]
+
+
+def _populate_creds_file():
+    """Prompt user for client ID and secret and save them."""
+    while True:
+        domain = click.prompt(
+            "Domain", default="services.sherlockml.com", err=True
```
Comment (acroz) [inline faculty/cli/cli.py]: Update this to `services.faculty.ai` (or whatever the services domain for cloud is going to be)?

File: faculty/cli/version.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+try:
+    __version__ = get_distribution("faculty").version
+except DistributionNotFound:
+    # package is not installed
+    pass
+
+
+def user_agent():
+    """Return user agent string."""
+    return "sml/{} (Python {}, {})".format(
```
[the remark this answers is not in the document]
Comment (srstevenson) [reply faculty/cli/version.py]: `faculty` sounds good to me.

File: faculty/cli/update.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+def _check_for_new_release():
+    current = StrictVersion(faculty.cli.version.__version__)
+    latest = max(_get_pypi_versions())
+    if current < latest:
+        template = (
+            "You are using faculty version {}, however version {} is "
+            "available.\n"
+            "You should upgrade with 'pip install --upgrade faculty'."
+        )
```
[the remark this answers is not in the document]
Comment (srstevenson) [reply faculty/cli/update.py]: I'm against folding `sml` into this library, so prefer having a separate `faculty-cli` package as a long term solution. As Andrew says, this avoids CLI users being told to upgrade when there is no functional change to the CLI, and additionally avoids library users from having an additional executable placed on their path when they're only using the Python API.

File: faculty/cli/config.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+
+_PROFILE_CACHE = None
+
+
+def _profile_name_from_faculty_env():
+    profile_name = os.getenv("FACULTY_ENV") or os.getenv("SHERLOCKML_ENV")
+    if profile_name is not None:
+        click.secho(
+            "FACULTY_ENV is deprecated and will be removed in a future "
```
[the remark this answers is not in the document]
Comment (srstevenson) [reply faculty/cli/config.py]: We've never used `FACULTY_ENV` before, so it doesn't seem worthwhile to add it here as it'll be deprecated from inception.

File: faculty/cli/cli.py (hunk 5 above)
Comment (srstevenson) [reply faculty/cli/cli.py]: `services.cloud.my.faculty.ai` says @neb42.

File: faculty/cli/config.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (zblz) [reply faculty/cli/config.py]: Indeed, I think the name change is a great moment to drop legacy stuff.

File: faculty/cli/update.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (zblz) [reply faculty/cli/update.py]: I was initially in favour of folding the CLI into `faculty`, but I'm coming around to a separate package being a better idea.

**Comment by mociarain** [comment]
This was closed as we no longer want to integrate the cli directly in this library. It has been moved [faculty-cli](https://pypi.org/project/faculty-cli/).

**Pull request closed without merging**